### PR TITLE
check the arguments number of `URLSearchParams`

### PR DESCRIPTION
**Pull request by justjavac**
`URLSearchParams` did not check the number of arguments passed, thereby causing some bugs.

When we use TypeScript, the tsc compiler will do parameter checking for us, but many times we write code in JavaScript.

```ts
let s = new URLSearchParams()
s.append()

// before
s.toString() === "undefined=undefined"

// after fix
// TypeError: URLSearchParams.append requires at least 2 arguments, but only 0 present
```

-------------

In addition, the first argument should be converted to `string` type

```ts
let s = new URLSearchParams()

// the `String` wrapper
s.set(new String("foo"), "bar")

// When calling `toString`, the keys and values are converted to strings
s.toString() === "foo=bar"

// but...

// before
s.has('foo') === false
s.get('foo') === null

// after fix
s.has('foo') === true
s.get('foo') === "bar"
```

### CODE REVIEW

**Comment by justjavac** [comment]
another question: should the private field `params` in `URLSearchParams` be moved outside the class, inside the module?

The private fields/methods in TypeScript will only be checked by the compiler and then compiled into a normal js field.

in js file:

```
// @ts-nocheck
s.set('foo', 'bar')
console.log(s.params)
```

will output `[ [ "foo", "bar" ] ]`.

File: js/url_search_params.ts [hunk 1]
```diff
@@ -112,22 +156,32 @@ export class URLSearchParams {
    *       searchParams.set('name', 'value');
    */
   set(name: string, value: string): void {
+    if (arguments.length < 2) {
+      // tslint:disable-next-line:max-line-length
+      const errMsg = `URLSearchParams.set requires at least 2 arguments, but only ${
+        arguments.length
+      } present`;
+      throw new TypeError(errMsg);
+    }
```
Comment (ry) [inline js/url_search_params.ts]: This bit of code is repeated quite often. Can it be abstracted into a separate function?

Comment (kyranet) [reply js/url_search_params.ts]: Perhaps something like this in util.ts:

```typescript
function requiredArguments(name: string, length: number, required: number): void {
  if (length < required) {
    const errMsg = `${name} requires at least ${required} argument${
      required === 1 ? '' : 's'}, but only ${length} present`;
    throw new TypeError(errMsg);
  }
}
```

And then calling ``requiredArguments('URLSearchParams.set', arguments.length, 2);``

**Comment by justjavac** [comment]
yes, I tested in 3 browsers and Node.

```js
let s = new URLSearchParams()
s.append()

// Firefox 66.0(a1)
// TypeError: URLSearchParams.append requires at least 2 arguments, but only 0 were passed

// Chrome 73.0.3649.0
// TypeError: Failed to execute 'append' on 'URLSearchParams': 2 arguments required, but only 0 present.

// Edge 17.17134
// 参数是必选项 
// Edge uses localized messages. I don't have English version of the browser, can only translate the general meaning: "Parameter is Required".

// Node 10.13.0
// TypeError [ERR_MISSING_ARGS]: The "name" and "value" arguments must be specified

// --------------------------

s.set(new String("foo"), "bar")

// Firefox, Chrome, Edge, Node
s.toString()    // "foo=bar"
s.has("foo")    // true
s.get("foo")    // "bar"
```

In the current implementation of deno, we use TypeScript‘s private field `params`, and Node.js uses `Symbol`. Although `Symbol` can create a unique key to prevent external direct access to its value, we can still get it through `getOwnPropertySymbols`:

```js
// Firefox, Chrome, Edge
Object.getOwnPropertySymbols(s)    // []

// Node
Object.getOwnPropertySymbols(s) 
// [ Symbol(query), Symbol(context) ]

const [ query, context ] = Object.getOwnPropertySymbols(s)
for (let k of Object.getOwnPropertySymbols(s)) {
    console.log(k, s[k])
}
s[query]    // [ 'foo', 'bar' ]
s[context]  // null

// Deno
s.params
// [ [ "foo", "bar" ] ]
```

Node uses a flat array, and Deno uses a two-dimensional array. Node is more graceful, but not compatible with the browser. If Deno wants to be compatible with the browser, then we can only move `params` outside `URLSearchParams`, inside the module, until V8 implements [private class fields](https://www.chromestatus.com/features/6035156464828416).

Maybe we can not care about private class field `params` right now(if there is no security risk for access to private field), we should just wait for TypeScript [Implement private fields proposal #9950](https://github.com/Microsoft/TypeScript/issues/9950)

**Pull request merged**